fe/fe-core/src/main/java/com/starrocks/qe/scheduler/dag/AllAtOnceExecutionSchedule.java

```diff
@@ -19,16 +19,57 @@
 import com.starrocks.qe.scheduler.Deployer;
 import com.starrocks.qe.scheduler.slot.DeployState;
 import com.starrocks.rpc.RpcException;
+import com.starrocks.server.GlobalStateMgr;
 import com.starrocks.thrift.TUniqueId;
 
 import java.util.ArrayList;
 import java.util.List;
+import java.util.concurrent.ExecutorService;
 
 // all at once execution schedule only schedule once.
 public class AllAtOnceExecutionSchedule implements ExecutionSchedule {
     private Coordinator coordinator;
     private Deployer deployer;
     private ExecutionDAG dag;
+    private volatile boolean cancelled = false;
+
+    class DeployMoreScanRangesTask implements Runnable {
+        List<DeployState> states;
+        private ExecutorService executorService;
+
+        DeployMoreScanRangesTask(List<DeployState> states, ExecutorService executorService) {
+            this.states = states;
+            this.executorService = executorService;
+        }
+
+        @Override
+        public void run() {
+            if (cancelled) {
+                return;
+            }
+            try {
+                states = coordinator.assignIncrementalScanRangesToDeployStates(deployer, states);
+                if (states.isEmpty()) {
+                    return;
+                }
+                for (DeployState state : states) {
+                    deployer.deployFragments(state);
+                }
+            } catch (StarRocksException | RpcException e) {
+                throw new RuntimeException(e);
+            }
+            // jvm should use tail optimization.
+            start();
+        }
```

Comment:
why not just use while(true)

Reply:
because there could be a lot of queries. it's better to give every query a chance to deliver some scan ranges.